File: demo_kinematics.py
"""运动学 demo —— Motor (SE(3) versor) 的直接应用, 无矩阵/四元数换算层。

三个机构同场景:
  1. 齿轮副: 齿数 16:8 → 角速度比精确 −1:2 (节圆相切, 相位错半齿);
  2. 曲柄滑块: 连杆 pose 每帧由两端点解算 (轴角转子, 无矩阵分解);
  3. 螺旋轨迹: M(s) = M₀·exp(s·log(M₀⁻¹M₁)) 插值同一 Motor 驱动小球
     (变换与几何同构 —— 轨迹本身是 motor 的解析函数)。

输出: artifacts/kinematics.gif (默认 72 帧)
用法: .venv/bin/python demo_kinematics.py [帧数]
"""

import math
import sys

from PIL import Image

from cga.engine import (
    AmbientLight,
    BoxGeometry,
    Color,
    CylinderGeometry,
    DirectionalLight,
    Mesh,
    MeshStandardMaterial,
    PerspectiveCamera,
    PlaneGeometry,
    PointLight,
    Renderer,
    Scene,
    SphereGeometry,
)
from cga.motors import Motor

# ── 构件 ──────────────────────────────────────────────────────────


def gear_meshes(r_hub, r_tooth, n_teeth, thick, color):
    """齿轮 (局部, 轴 +Y): 轮毂圆柱 + 径向齿阵列 (Motor 复合定位)。"""
    mat = MeshStandardMaterial(color, roughness=0.35, metalness=0.7)
    out = [
        Mesh(
            CylinderGeometry(r_hub, thick),
            mat,
            rotation_axis=(1, 0, 0),
            rotation_angle=-math.pi / 2,  # 局部 +Z → +Y
        )
    ]
    r_mid = r_hub + (r_tooth - r_hub) / 2
    for i in range(n_teeth):
        a = i * 2 * math.pi / n_teeth
        # 齿: 先平移到节圆, 再绕 +Y 转 (T·R 顺序: 右先作用)
        m = Motor.rotor((0, 1, 0), a).gp(Motor.translator((r_mid, 0, 0)))
        out.append(Mesh(BoxGeometry(r_tooth - r_hub + 0.06, thick, 0.16), mat, motor=m))
    return out


def frame_motor(axis, angle, point=(0.0, 0.0, 0.0)):
    """绕任意过点轴的帧变换: M = T(p)·R(axis, angle)·T(−p)。"""
    return (
        Motor.translator(point)
        .gp(Motor.rotor(axis, angle))
        .gp(Motor.translator(tuple(-v for v in point)))
    )


def rod_motor(p, q):
    """连杆 (局部 +Z 圆柱, 居中) → 从 p 到 q 的刚体 pose 与长度。"""
    d = [q[i] - p[i] for i in range(3)]
    length = math.sqrt(sum(v * v for v in d))
    u = [v / length for v in d]
    # 转子: +Z → u (轴角, 无矩阵分解)
    ax = (-u[1], u[0])  # ẑ×u = (−uy, ux, 0)
    an = math.sqrt(ax[0] ** 2 + ax[1] ** 2)
    if an < 1e-12:
        rot = Motor.rotor((1, 0, 0), 0.0 if u[2] > 0 else math.pi)
    else:
        angle = math.acos(max(-1.0, min(1.0, u[2])))
        rot = Motor.rotor((ax[0] / an, ax[1] / an, 0.0), angle)
    mid = tuple((p[i] + q[i]) / 2 for i in range(3))
    return Motor.translator(mid).gp(rot), length


# ── 场景 ──────────────────────────────────────────────────────────

STEEL = MeshStandardMaterial(Color(0x9BA1A6), roughness=0.35, metalness=0.75)
BRASS = MeshStandardMaterial(Color(0xC8A24A), roughness=0.3, metalness=0.8)
DARK = MeshStandardMaterial(Color(0x4A4F54), roughness=0.5, metalness=0.6)
CRANK_C = (-1.6, 0.6, 2.6)  # 曲柄中心
ROD_LEN = 1.35


def build_scene():
    scene = Scene()
    scene.add(
        Mesh(
            PlaneGeometry((0, 1, 0), -0.05),
            MeshStandardMaterial(Color(0x3A4046), roughness=0.9),
        ),
        DirectionalLight(intensity=0.5, direction=(0.4, 1.0, 0.5)),
        PointLight(intensity=0.5, position=(-4, 6, 4)),
        AmbientLight(intensity=0.4),
    )
    # 齿轮副 (节圆相切于 x 轴): 大轮 16 齿在原点, 小轮 8 齿在 x=2.4
    big = gear_meshes(1.28, 1.6, 16, 0.4, Color(0xC8A24A))
    small = gear_meshes(0.64, 0.8, 8, 0.4, Color(0x9BA1A6))
    for m in small:
        m.motor_override = Motor.translator((2.4, 0, 0)).gp(m.motor())
    for m in big + small:
        scene.add(m)
    # 曲柄滑块 (后方 z=2.6 平面): 曲柄盘 + 连杆 + 滑块 + 导轨
    scene.add(
        Mesh(
            CylinderGeometry(0.55, 0.25),
            DARK,
            position=CRANK_C,
            rotation_axis=(1, 0, 0),
            rotation_angle=-math.pi / 2,
        )
    )
    rod = Mesh(CylinderGeometry(0.09, 1.0), BRASS)  # 每帧重设 pose/长度
    slider = Mesh(BoxGeometry(0.5, 0.4, 0.35), STEEL)
    scene.add(rod, slider)
    scene.add(Mesh(BoxGeometry(2.6, 0.08, 0.5), DARK, position=(0.0, 0.4, 2.6)))
    # 螺旋轨迹小球 + 轨道参照珠 (motor exp/log 解析轨迹, 避开机构上方)
    ball = Mesh(
        SphereGeometry(0.18), MeshStandardMaterial(Color(0xC0392B), roughness=0.3)
    )
    scene.add(ball)
    # 轨迹端点: M0 在左侧低位, M1 在右侧高位并绕 +Y 转 3π/2
    m0 = Motor.translator((-3.2, 0.5, 2.2))
    m1 = Motor.translator((3.6, 1.8, 3.4)).gp(Motor.rotor((0, 1, 0), 1.5 * math.pi))
    twist = m0.inverse().gp(m1).log()
    for k in range(9):
        scene.add(
            Mesh(
                SphereGeometry(0.05),
                MeshStandardMaterial(Color(0x7F8C8D), roughness=0.6),
                motor=m0.gp(Motor.exp(twist, k / 8)),
            )
        )
    return scene, big, small, rod, slider, ball, (m0, twist)


def main():
    n_frames = int(sys.argv[1]) if len(sys.argv) > 1 else 72
    scene, big, small, rod, slider, ball, (m0, twist) = build_scene()
    big_local = [m.motor() for m in big]  # 局部 motor (每帧左乘机构变换)
    small_local = [m.motor() for m in small]
    cam = PerspectiveCamera(
        fov=48, aspect=4 / 3, position=(0.4, 5.6, 8.8), target=(0.4, 0.2, 1.2)
    )
    cam.look_at((0.4, 0.2, 1.2))
    renderer = Renderer(480, 360, aa=2)
    frames = []
    for f in range(n_frames):
        s = f / n_frames
        # 齿轮: 大轮 ω = 2π/圈, 小轮 −2ω (齿数比 16:8) + 半齿相位
        g1 = frame_motor((0, 1, 0), 2 * math.pi * s)
        for m, lm in zip(big, big_local):
            m.motor_override = g1.gp(lm)
        g2 = frame_motor((0, 1, 0), -4 * math.pi * s + math.pi / 8, point=(2.4, 0, 0))
        for m, lm in zip(small, small_local):
            m.motor_override = g2.gp(lm)
        # 曲柄滑块: 销 P 在半径 0.35 圆上, 滑块约束在 x 方向导轨
        th = 2 * math.pi * s
        p = (
            CRANK_C[0] + 0.35 * math.cos(th),
            CRANK_C[1],
            CRANK_C[2] + 0.35 * math.sin(th),
        )
        xs = p[0] + math.sqrt(ROD_LEN**2 - (CRANK_C[2] - p[2]) ** 2)
        q = (xs, CRANK_C[1], CRANK_C[2])
        rm, rl = rod_motor(p, q)
        rod.motor_override = rm
        rod.geometry = CylinderGeometry(0.09, rl)
        slider.position = q
        slider.motor_override = None
        # 螺旋小球: M(s) = M0·exp(s·log(M0⁻¹M1))
        ball.motor_override = m0.gp(Motor.exp(twist, s))
        img = renderer.render(scene, cam)
        frames.append(
            Image.frombytes("RGBA", (480, 360), Renderer.frame_to_bytes(img)).convert(
                "P"
            )
        )
        if (f + 1) % 12 == 0:
            print(f"frame {f + 1}/{n_frames}")
    frames[0].save(
        "artifacts/kinematics.gif",
        save_all=True,
        append_images=frames[1:],
        duration=1000 // 24,
        loop=0,
    )
    print("saved artifacts/kinematics.gif")


if __name__ == "__main__":
    main()

File: crates/cga-examples/src/bin/demo_kinematics.rs
// Kinematics demo: gears + crank-slider + spiral trajectory driven by Motor.
// Saves PNG frames and an animated GIF (assembled in V).
use cga_core::*;
use cga_examples::{data_f32, mlx_frame_gc};
use cga_gpu::*;
use std::f64::consts::PI;

struct KinScene {
    scene: Scene,
    big: Vec<usize>,
    small: Vec<usize>,
    rod: usize,
    slider: usize,
    ball: usize,
    m0: Multivector,
    twist: Multivector,
}

fn gear_meshes(
    ks: &mut KinScene,
    r_hub: f64,
    r_tooth: f64,
    n_teeth: i32,
    thick: f64,
    color: Color,
) -> Vec<usize> {
    let mat = standard_material(MaterialParams {
        color,
        roughness: 0.35,
        metalness: 0.7,
        emissive: color_hex(0x000000),
        opacity: 1.0,
        ior: 1.5,
        absorption: 0.0,
    });
    let mut idx: Vec<usize> = Vec::new();
    ks.scene.add_mesh(mesh(MeshParams {
        geometry: Geometry::CylinderGeometry(cylinder_geometry(r_hub, thick)),
        material: mat.clone(),
        position: [0.0, 0.0, 0.0],
        rotation_axis: [1.0, 0.0, 0.0],
        rotation_angle: -PI / 2.0,
        motor: None,
    }));
    idx.push(ks.scene.objects.len() - 1);
    let r_mid = r_hub + (r_tooth - r_hub) / 2.0;
    for i in 0..n_teeth {
        let a = f64::from(i) * 2.0 * PI / f64::from(n_teeth);
        let m = motor_rotor([0.0, 1.0, 0.0], a).gp(&translator([r_mid, 0.0, 0.0]));
        ks.scene.add_mesh(mesh(MeshParams {
            geometry: Geometry::BoxGeometry(box_geometry(r_tooth - r_hub + 0.06, thick, 0.16)),
            material: mat.clone(),
            position: [0.0, 0.0, 0.0],
            rotation_axis: [0.0, 0.0, 1.0],
            rotation_angle: 0.0,
            motor: Some(m),
        }));
        idx.push(ks.scene.objects.len() - 1);
    }
    idx
}

fn frame_motor(axis: [f64; 3], angle: f64, point: [f64; 3]) -> Multivector {
    translator(point)
        .gp(&motor_rotor(axis, angle))
        .gp(&translator([-point[0], -point[1], -point[2]]))
}

fn rod_motor(p: [f64; 3], q: [f64; 3]) -> (Multivector, f64) {
    let d = [q[0] - p[0], q[1] - p[1], q[2] - p[2]];
    let length = (d[0] * d[0] + d[1] * d[1] + d[2] * d[2]).sqrt();
    let u = [d[0] / length, d[1] / length, d[2] / length];
    let ax = [-u[1], u[0]];
    let an = (ax[0] * ax[0] + ax[1] * ax[1]).sqrt();
    let rot = if an < 1e-12 {
        motor_rotor([1.0, 0.0, 0.0], if u[2] > 0.0 { 0.0 } else { PI })
    } else {
        let angle = u[2].clamp(-1.0, 1.0).acos();
        motor_rotor([ax[0] / an, ax[1] / an, 0.0], angle)
    };
    let mid = [
        (p[0] + q[0]) / 2.0,
        (p[1] + q[1]) / 2.0,
        (p[2] + q[2]) / 2.0,
    ];
    (translator(mid).gp(&rot), length)
}

fn build_scene() -> KinScene {
    let mut ks = KinScene {
        scene: scene(None),
        big: Vec::new(),
        small: Vec::new(),
        rod: 0,
        slider: 0,
        ball: 0,
        m0: motor_identity(),
        twist: motor_identity(),
    };
    ks.scene.add_mesh(mesh(MeshParams {
        geometry: Geometry::PlaneGeometry(plane_geometry([0.0, 1.0, 0.0], -0.05)),
        material: standard_material(MaterialParams {
            color: color_hex(0x3A4046),
            roughness: 0.9,
            metalness: 0.0,
            emissive: color_hex(0x000000),
            opacity: 1.0,
            ior: 1.5,
            absorption: 0.0,
        }),
        position: [0.0, 0.0, 0.0],
        rotation_axis: [0.0, 0.0, 1.0],
        rotation_angle: 0.0,
        motor: None,
    }));
    ks.scene
        .add_light(directional_light(color_hex(0xFFFFFF), 0.5, [0.4, 1.0, 0.5]));
    ks.scene
        .add_light(point_light(color_hex(0xFFFFFF), 0.5, [-4.0, 6.0, 4.0]));
    ks.scene.add_light(ambient_light(color_hex(0xFFFFFF), 0.4));
    // gears: big (16 teeth) at origin, small (8 teeth) at x=2.4
    ks.big = gear_meshes(&mut ks, 1.28, 1.6, 16, 0.4, color_hex(0xC8A24A));
    ks.small = gear_meshes(&mut ks, 0.64, 0.8, 8, 0.4, color_hex(0x9BA1A6));
    for &idx in &ks.small.clone() {
        let m = translator([2.4, 0.0, 0.0]).gp(&ks.scene.objects[idx].motor());
        ks.scene.objects[idx].motor_override = Some(m);
    }
    // crank-slider (rear z=2.6 plane)
    let crank_c = [-1.6, 0.6, 2.6];
    ks.scene.add_mesh(mesh(MeshParams {
        geometry: Geometry::CylinderGeometry(cylinder_geometry(0.55, 0.25)),
        material: standard_material(MaterialParams {
            color: color_hex(0x4A4F54),
            roughness: 0.5,
            metalness: 0.6,
            emissive: color_hex(0x000000),
            opacity: 1.0,
            ior: 1.5,
            absorption: 0.0,
        }),
        position: crank_c,
        rotation_axis: [1.0, 0.0, 0.0],
        rotation_angle: -PI / 2.0,
        motor: None,
    }));
    ks.scene.add_mesh(mesh(MeshParams {
        geometry: Geometry::CylinderGeometry(cylinder_geometry(0.09, 1.0)),
        material: standard_material(MaterialParams {
            color: color_hex(0xC8A24A),
            roughness: 0.3,
            metalness: 0.8,
            emissive: color_hex(0x000000),
            opacity: 1.0,
            ior: 1.5,
            absorption: 0.0,
        }),
        position: [0.0, 0.0, 0.0],
        rotation_axis: [0.0, 0.0, 1.0],
        rotation_angle: 0.0,
        motor: None,
    }));
    ks.rod = ks.scene.objects.len() - 1;
    ks.scene.add_mesh(mesh(MeshParams {
        geometry: Geometry::BoxGeometry(box_geometry(0.5, 0.4, 0.35)),
        material: standard_material(MaterialParams {
            color: color_hex(0x9BA1A6),
            roughness: 0.35,
            metalness: 0.75,
            emissive: color_hex(0x000000),
            opacity: 1.0,
            ior: 1.5,
            absorption: 0.0,
        }),
        position: [0.0, 0.0, 0.0],
        rotation_axis: [0.0, 0.0, 1.0],
        rotation_angle: 0.0,
        motor: None,
    }));
    ks.slider = ks.scene.objects.len() - 1;
    ks.scene.add_mesh(mesh(MeshParams {
        geometry: Geometry::BoxGeometry(box_geometry(2.6, 0.08, 0.5)),
        material: standard_material(MaterialParams {
            color: color_hex(0x4A4F54),
            roughness: 0.5,
            metalness: 0.6,
            emissive: color_hex(0x000000),
            opacity: 1.0,
            ior: 1.5,
            absorption: 0.0,
        }),
        position: [0.0, 0.4, 2.6],
        rotation_axis: [0.0, 0.0, 1.0],
        rotation_angle: 0.0,
        motor: None,
    }));
    // spiral ball
    ks.scene.add_mesh(mesh(MeshParams {
        geometry: Geometry::SphereGeometry(sphere_geometry(0.18)),
        material: standard_material(MaterialParams {
            color: color_hex(0xC0392B),
            roughness: 0.3,
            metalness: 0.0,
            emissive: color_hex(0x000000),
            opacity: 1.0,
            ior: 1.5,
            absorption: 0.0,
        }),
        position: [0.0, 0.0, 0.0],
        rotation_axis: [0.0, 0.0, 1.0],
        rotation_angle: 0.0,
        motor: None,
    }));
    ks.ball = ks.scene.objects.len() - 1;
    let m0 = translator([-3.2, 0.5, 2.2]);
    let m1 = translator([3.6, 1.8, 3.4]).gp(&motor_rotor([0.0, 1.0, 0.0], 1.5 * PI));
    ks.m0 = m0;
    ks.twist = m0.reverse().gp(&m1).log();
    for k in 0..9 {
        ks.scene.add_mesh(mesh(MeshParams {
            geometry: Geometry::SphereGeometry(sphere_geometry(0.05)),
            material: standard_material(MaterialParams {
                color: color_hex(0x7F8C8D),
                roughness: 0.6,
                metalness: 0.0,
                emissive: color_hex(0x000000),
                opacity: 1.0,
                ior: 1.5,
                absorption: 0.0,
            }),
            position: [0.0, 0.0, 0.0],
            rotation_axis: [0.0, 0.0, 1.0],
            rotation_angle: 0.0,
            motor: Some(m0.gp(&motor_exp(&ks.twist, f64::from(k) / 8.0))),
        }));
    }
    ks
}

fn main() {
    let mut ks = build_scene();
    let n_frames = 24;
    let mut big_local: Vec<Multivector> = Vec::new();
    for &idx in &ks.big {
        big_local.push(ks.scene.objects[idx].motor());
    }
    let mut small_local: Vec<Multivector> = Vec::new();
    for &idx in &ks.small {
        small_local.push(ks.scene.objects[idx].motor());
    }
    let mut cam = perspective_camera(
        48.0,
        4.0 / 3.0,
        0.1,
        100.0,
        [0.4, 5.6, 8.8],
        [0.4, 0.2, 1.2],
        [0.0, 1.0, 0.0],
    );
    cam.look_at([0.4, 0.2, 1.2], None);
    let mut r = renderer(360, 270, 2, 3);
    let mut gif_frames: Vec<Vec<u8>> = Vec::new();
    let big = ks.big.clone();
    let small = ks.small.clone();
    for f in 0..n_frames {
        let s = f64::from(f) / f64::from(n_frames);
        let g1 = frame_motor([0.0, 1.0, 0.0], 2.0 * PI * s, [0.0, 0.0, 0.0]);
        for (i, &idx) in big.iter().enumerate() {
            ks.scene.objects[idx].motor_override = Some(g1.gp(&big_local[i]));
        }
        let g2 = frame_motor([0.0, 1.0, 0.0], -4.0 * PI * s + PI / 8.0, [2.4, 0.0, 0.0]);
        for (i, &idx) in small.iter().enumerate() {
            ks.scene.objects[idx].motor_override = Some(g2.gp(&small_local[i]));
        }
        let th = 2.0 * PI * s;
        let p = [-1.6 + 0.35 * th.cos(), 0.6, 2.6 + 0.35 * th.sin()];
        let xs = p[0] + (1.35 * 1.35 - (2.6 - p[2]) * (2.6 - p[2])).sqrt();
        let q = [xs, 0.6, 2.6];
        let (rm, rl) = rod_motor(p, q);
        ks.scene.objects[ks.rod].motor_override = Some(rm);
        ks.scene.objects[ks.rod].geometry = Geometry::CylinderGeometry(cylinder_geometry(0.09, rl));
        ks.scene.objects[ks.slider].position = q;
        ks.scene.objects[ks.slider].motor_override = None;
        ks.scene.objects[ks.ball].motor_override = Some(ks.m0.gp(&motor_exp(&ks.twist, s)));
        let img = r.render(ks.scene.clone(), cam);
        gif_frames.push(f32_rgba_to_u8(&data_f32(&img)));
        drop(img);
        // 每帧收尾：释放死 Metal buffer（Rust drop），再把 MLX 不复用的
        // cache 还给 OS（见 editor/server.v；V 里是 gc_collect + clear_cache）
        mlx_frame_gc();
    }
    // V: os.dir(@FILE) — Rust 版以仓库根目录为 CWD 运行
    let out = "examples/kinematics";
    save_gif(&format!("{out}/kinematics.gif"), &gif_frames, 360, 270, 5);
}
